### ISSUE TITLE
Inconsistent finalizer test behavior on Ubuntu

### ISSUE THREAD

**Issue by swgillespie** [dotnet/coreclr#4093]
See http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_tst_prtest/501/testReport/junit/GC.Scenarios.DoublinkList/dlcollect/dlcollect/ for a failure. I failed to get a repro after running this for about a week last week, but it happened again in CI.
- http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_tst_prtest/545/testReport/GC.Scenarios.DoublinkList/dlcollect/dlcollect/history/
- http://dotnet-ci.cloudapp.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_tst_prtest/545/testReport/GC.Scenarios.LeakWheel/leakwheel/leakwheel/history/

**Comment by swgillespie** [comment]
Spoke to @Maoni0 about this one - I have a repro but since it only repros on checked Ubuntu builds it's not very high priority. Moving to 1.1.0.

**Comment by jkotas** [comment]
#7558 - GC.API.GCHandleCollector.Usage.Usage test failed with "Case 3 failed: threshold not increasing!" on Ubuntu

**Comment by swgillespie** [comment]
@jkotas that's interesting, never seen that one before:

```
09:41:43 FAILED   - GC/API/GCHandleCollector/Usage/Usage.sh
09:41:43                BEGIN EXECUTION
09:41:43                /mnt/j/workspace/dotnet_coreclr/master/checked_ubuntu_tst_prtest/bin/tests/Windows_NT.x64.Checked/Tests/coreoverlay/corerun Usage.exe
09:41:43                Case 1 Passed!
09:41:43                100, 100
09:41:43                Case 2 Passed!
09:41:43                Case 3 failed: threshold not increasing!
09:41:43                
09:41:43                Test Failed!
09:41:43                Expected: 100
09:41:43                Actual: 1
09:41:43                END EXECUTION - FAILED
```

**Comment by jkotas** [comment]
It was seen before in #4400 that was resolved as duplicate of this issue.

**Comment by swgillespie** [comment]
ah, you're right - sorry for my lapse in memory. I suppose it's much less common than the other manifestations.

**Comment by AndyAyersMS** [comment]
Here's another instance with a crash dump... note this is from an unmerged PR so it might be something I did. But it might not be since earlier versions of this that are very similar did not hit this problem.

https://ci.dot.net/job/dotnet_coreclr/job/master/job/checked_ubuntu_flow_prtest/1183/

**Comment by swgillespie** [comment]
cc @sergiy-k @Maoni0 

I took a look at this in-depth today and I can consistently reproduce the test failure by using a debugger to set up this sequence of events:

1. A GC occurs (in this case, the test calls `GC.Collect()`
2. GC calls `GCToEEInterface::EnableFinalization`, waking up the finalizer thread
3. Finalizer thread begins running finalizers asynchronously
4. At the same time, main thread qcalls into `GCInterface::WaitForPendingFinalizers`, which in turn calls `FinalizerThread::FinalizerThreadWait`
5. After suspending the finalizer thread [here](https://github.com/dotnet/coreclr/blob/master/src/vm/finalizerthread.cpp#L683) in a debugger and allowing the main thread to reach its wait on `hEventFinalizerDone` before the finalizer thread sets it, the finalizer thread sets the event, waking up the main thread without having run any finalizers since the last pass.

This race means that this code:
```csharp
GC.Collect();
GC.WaitForPendingFinalizers();
GC.Collect();
AssertFinalizersHaveRun();
```

Is inherently racy, since this sequence of events is possible
1. A GC occurs (triggered through allocation)
2. `GC.Collect()` is called, does a GC, and signals that there are finalizers to be run, but the finalizer thread has still not signaled that it is done finalizing things from the first GC
3. `GC.WaitForPendingFinalizers()` catches the finalizer thread before it signals that it is done making
a single pass, so it returns immediately
4. `GC.Collect()` is called, does another GC, realizes that almost everything is finalizer-reachable and doesn't do much
5. `AssertFinalizersHaveRun()` races with the finalizer thread - the finalizer thread is finalizing things while the main thread is asserting that everything has been finalized.

This race is apparently known and considered acceptable: https://github.com/dotnet/coreclr/blob/master/src/vm/finalizerthread.cpp#L685-L688 - so I suppose what's left to do is figure out how to harden these tests against this race. Since we don't really make any guarantees about when things get finalized, I can't think of anything off the top of my head more robust than simply sleeping for some period of time before asserting that finalizers have run.

**Comment by jkotas** [comment]
> I can't think of anything off the top of my head more robust than simply sleeping for some period of time

You won't be really testing much if you add sleep. If WaitForPendingFinalizers just returns without waiting for anything, the tests hardened by the extra sleep will pass.

I think you may want to call WaitForPendingFinalizers finalizers twice to make it robust. It should ensure that the finalizer went through the whole finalization queue at least once. Or maybe even better - change the implementation of `GCInterface::WaitForPendingFinalizers` to call `FinalizerThread::FinalizerThreadWait` twice that should fix the race in the API.

**Comment by Maoni0** [comment]
the root problem here really is just the last GC.Collect added more things to the FQ. so the test can just call this whole sequence (Collect/Wait/Collect) in a  loop and check to see if we make progress (since it knows exactly how many objects it expect to be finalized) in each iteration; if we do then do another iteration. Otherwise we know we have a real issue. For this particular test calling it twice would solve the problem as it's one off; but it could easily be the case that another test will not be solved by calling it twice (if the 2nd iter again added stuff to the finalize queue).

**Comment by swgillespie** [comment]
I noticed some tests already take some measures against this race:
* https://github.com/dotnet/coreclr/blob/master/tests/src/GC/Scenarios/LeakWheel/leakwheel.cs#L149
* https://github.com/dotnet/coreclr/blob/master/tests/src/GC/Scenarios/DoublinkList/dlstack.cs#L79
* https://github.com/dotnet/coreclr/blob/master/tests/src/GC/Scenarios/DoublinkList/dlcollect.cs#L71

Tomorrow I will convert tests using these patterns to the Collect/Wait/Collect loop per Maoni and Jan's comments. It'll be useful to do the same thing in each of these tests, I think.
